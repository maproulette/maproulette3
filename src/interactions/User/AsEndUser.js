import { GUEST_USER_ID } from '../../services/User/User'
import { ROLE_SUPERUSER } from '../../services/Grant/Role'
import _find from 'lodash/find'
import _filter from 'lodash/filter'
import _isObject from 'lodash/isObject'
import _isNumber from 'lodash/isNumber'

/**
 * Provides basic methods for interacting with users.
 */
export class AsEndUser {
  constructor(user) {
    this.user = user
  }

  /**
   * Returns true if the user is logged in, false otherwise.
   */
  isLoggedIn() {
    return _isObject(this.user) &&
           _isNumber(this.user.id) && this.user.id !== GUEST_USER_ID
  }

  /**
   * Returns true if the user is a super user, false otherwise.
   */
  isSuperUser() {
    return this.isLoggedIn() &&
           !!_find(this.user.grants, {role: ROLE_SUPERUSER})
  }

  /**
   * Returns true if the user is a reviewer.
   */
  isReviewer() {
    return this.isLoggedIn() && this.user.settings.isReviewer
  }

  /**
   * Returns true if the user's work needs to be reviewed.
   */
  needsReview() {
    return this.isLoggedIn() && this.user.settings.needsReview
  }

  /**
   * Returns true if the user has at least one notification that is not marked
   * as read
   */
  hasUnreadNotifications() {
    return this.isLoggedIn() &&
           !!_find(this.user.notifications, {isRead: false})
  }

  /**
   * Returns the number of notifications not marked as read
   */
  unreadNotificationCount() {
    if (!this.isLoggedIn()) {
      return 0
    }

    return _filter(this.user.notifications, {isRead: false}).length
  }
}

export default user => new AsEndUser(user)
