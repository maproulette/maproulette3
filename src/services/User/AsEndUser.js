import { GUEST_USER_ID,
         SUPERUSER_GROUP_TYPE } from './User'
import _find from 'lodash/find'
import _isObject from 'lodash/isObject'
import _isNumber from 'lodash/isNumber'

/**
 * Provides basic methods for interacting with users.
 */
export default class AsEndUser {
  constructor(user) {
    this.user = user
  }

  /**
   * Returns true if the user is logged in, false otherwise.
   */
  isLoggedIn() {
    return _isObject(this.user) &&
           _isNumber(this.user.id) && this.user.id !== GUEST_USER_ID
  }

  /**
   * Returns true if the user is a super user, false otherwise.
   */
  isSuperUser() {
    return this.isLoggedIn() &&
           !!_find(this.user.groups, {groupType: SUPERUSER_GROUP_TYPE})
  }
}
