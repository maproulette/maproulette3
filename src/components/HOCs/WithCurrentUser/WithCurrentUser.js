import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import { denormalize } from 'normalizr'
import _debounce from 'lodash/debounce'
import _get from 'lodash/get'
import { logoutUser,
         fetchUser,
         loadCompleteUser,
         saveChallenge, unsaveChallenge,
         saveTask, unsaveTask,
         fetchTopChallenges,
         fetchSavedChallenges,
         fetchSavedTasks,
         fetchUserActivity,
         updateUserAppSetting,
         userDenormalizationSchema } from '../../../services/User/User'
import AsEndUser from '../../../interactions/User/AsEndUser'

const APP_ID = "mr3Frontend"

/**
 * WithCurrentUser passes down the current user from the redux store.  If the
 * user is non-null, it also automatically adds isSignedIn and isSuperUser
 * fields to the user object (isSignedIn should be checked as the user could be
 * a guest user). Various functions are also made available for managing saved
 * user challenges and tasks, as well as logging out the user.
 *
 * @author [Neil Rotstan](https://github.com/nrotstan)
 */
const WithCurrentUser =
  WrappedComponent => connect(mapStateToProps, mapDispatchToProps)(WrappedComponent)

export const mapStateToProps = state => {
  const props = {user: null, allUsers: null}

  const userId = _get(state, 'currentUser.userId')
  const userEntity = _get(state, `entities.users.${userId}`)
  if (userEntity) {
    props.user =
      denormalize(userEntity, userDenormalizationSchema(), state.entities)

    if (props.user) {
      const endUser = AsEndUser(props.user)
      props.user.isLoggedIn = endUser.isLoggedIn()
      props.user.isSuperUser = endUser.isSuperUser()
      props.user.hasUnreadNotifications = endUser.hasUnreadNotifications()
    }
  }

  props.allUsers = _get(state, "entities.users")
  return props
}

export const mapDispatchToProps = dispatch => {
  const actions = bindActionCreators({
    fetchUser,
    loadCompleteUser,
    logoutUser,
    fetchSavedChallenges,
    saveChallenge,
    unsaveChallenge,
    fetchSavedTasks,
    saveTask,
    unsaveTask,
    fetchTopChallenges,
    fetchUserActivity,
  }, dispatch)

  actions.updateUserAppSetting = _debounce((userId, setting) => {
    return dispatch(updateUserAppSetting(userId, APP_ID, setting))
  }, 100)

  actions.getUserAppSetting = (user, settingName) => {
    return _get(user, `properties.${APP_ID}.settings.${settingName}`)
  }

  return actions
}

export default WithCurrentUser
