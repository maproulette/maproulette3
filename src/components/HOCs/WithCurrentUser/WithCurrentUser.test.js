import React, { Component } from 'react'
import { mapStateToProps, mapDispatchToProps } from './WithCurrentUser'
import AsEndUser from '../../../interactions/User/AsEndUser'
import { denormalize } from 'normalizr'
import { logoutUser,
         saveChallenge, unsaveChallenge,
         saveTask, unsaveTask,
         userDenormalizationSchema } from '../../../services/User/User'

jest.mock('normalizr')
jest.mock('../../../services/User/User')
jest.mock('../../../interactions/User/AsEndUser')

const schema = {foo: "bar"}
let basicState = null

denormalize.mockImplementation((user, schema, entities) => user)
userDenormalizationSchema.mockImplementation(() => schema)

const isLoggedIn = jest.fn()
const isSuperUser = jest.fn()
const hasUnreadNotifications = jest.fn()
const unreadNotificationCount = jest.fn()
AsEndUser.mockImplementation(() => ({
  isLoggedIn,
  isSuperUser,
  hasUnreadNotifications,
  unreadNotificationCount,
}))


beforeEach(() => {
  AsEndUser.mockClear()

  basicState = {
    currentUser: {
      userId: 456,
    },
    entities: {
      users: {
        123: {
          id: 123,
        },
        456: {
          id: 456,
        },
      }
    }
  }
})

test("mapStateToProps provides the current user in denormalized form", () => {
  const mappedProps = mapStateToProps(basicState)

  expect(denormalize).toBeCalledWith(basicState.entities.users[basicState.currentUser.userId],
                                     schema, basicState.entities)
  expect(mappedProps.user.id).toBe(basicState.currentUser.userId)

  expect(mappedProps).toMatchSnapshot()
})

test("mapStateToProps adds isLoggedIn and isSuperUser fields to the user", () => {
  isLoggedIn.mockReturnValueOnce(true)
  isSuperUser.mockReturnValueOnce(true)

  const mappedProps = mapStateToProps(basicState)

  expect(isLoggedIn).toHaveBeenCalled()
  expect(mappedProps.user.isLoggedIn).toBe(true)

  expect(isSuperUser).toHaveBeenCalled()
  expect(mappedProps.user.isSuperUser).toBe(true)

  expect(mappedProps).toMatchSnapshot()
})

test("maps the user to null if there is no current user", () => {
  basicState.currentUser = null

  const mappedProps = mapStateToProps(basicState)
  expect(mappedProps.user).toBe(null)
})

test("mapDispatchToProps makes the logoutUser() function available", () => {
  const dispatch = jest.fn()
  const mappedProps = mapDispatchToProps(dispatch)

  mappedProps.logoutUser()
  expect(dispatch).toBeCalled()
  expect(logoutUser).toBeCalled()
})

test("mapDispatchToProps makes the saveChallenge() function available", () => {
  const dispatch = jest.fn()
  const mappedProps = mapDispatchToProps(dispatch)

  const userId = 123
  const challengeId = 987

  mappedProps.saveChallenge(userId, challengeId)
  expect(dispatch).toBeCalled()
  expect(saveChallenge).toBeCalledWith(userId, challengeId)
})

test("mapDispatchToProps makes the unsaveChallenge() function available", () => {
  const dispatch = jest.fn()
  const mappedProps = mapDispatchToProps(dispatch)

  const userId = 123
  const challengeId = 987

  mappedProps.unsaveChallenge(userId, challengeId)
  expect(dispatch).toBeCalled()
  expect(unsaveChallenge).toBeCalledWith(userId, challengeId)
})

test("mapDispatchToProps makes the saveTask() function available", () => {
  const dispatch = jest.fn()
  const mappedProps = mapDispatchToProps(dispatch)

  const userId = 123
  const taskId = 987

  mappedProps.saveTask(userId, taskId)
  expect(dispatch).toBeCalled()
  expect(saveTask).toBeCalledWith(userId, taskId)
})

test("mapDispatchToProps makes the unsaveTask() function available", () => {
  const dispatch = jest.fn()
  const mappedProps = mapDispatchToProps(dispatch)

  const userId = 123
  const taskId = 987

  mappedProps.unsaveTask(userId, taskId)
  expect(dispatch).toBeCalled()
  expect(unsaveTask).toBeCalledWith(userId, taskId)
})
