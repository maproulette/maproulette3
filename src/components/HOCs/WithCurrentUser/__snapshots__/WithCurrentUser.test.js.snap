// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapStateToProps adds isLoggedIn and isSuperUser fields to the user 1`] = `
Object {
  "allUsers": Object {
    "123": Object {
      "id": 123,
    },
    "456": Object {
      "hasUnreadNotifications": undefined,
      "id": 456,
      "isLoggedIn": true,
      "isSuperUser": true,
      "unreadNotificationCount": undefined,
    },
  },
  "user": Object {
    "hasUnreadNotifications": undefined,
    "id": 456,
    "isLoggedIn": true,
    "isSuperUser": true,
    "unreadNotificationCount": undefined,
  },
}
`;

exports[`mapStateToProps provides the current user in denormalized form 1`] = `
Object {
  "allUsers": Object {
    "123": Object {
      "id": 123,
    },
    "456": Object {
      "hasUnreadNotifications": undefined,
      "id": 456,
      "isLoggedIn": undefined,
      "isSuperUser": undefined,
      "unreadNotificationCount": undefined,
    },
  },
  "user": Object {
    "hasUnreadNotifications": undefined,
    "id": 456,
    "isLoggedIn": undefined,
    "isSuperUser": undefined,
    "unreadNotificationCount": undefined,
  },
}
`;
